File: 프로그래머스/2/340211. ［PCCP 기출문제］ 3번 ／ 충돌위험 찾기/［PCCP 기출문제］ 3번 ／ 충돌위험 찾기.java
import java.util.*;

class Solution {
    class Robot {
        int x;
        int y;
        List<int[]> path = new ArrayList<>();
        boolean goal = false;

        public Robot(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    public int solution(int[][] points, int[][] routes) {
        int ans = 0;
        List<Robot> robots = new ArrayList<>();
        int[][] dir = {{1,0},{-1,0},{0,1},{0,-1}};

        // robots 초기화
        for (int i = 0; i < routes.length; i++) {
            int x = points[routes[i][0] - 1][0] - 1;
            int y = points[routes[i][0] - 1][1] - 1;

            Robot r = new Robot(x, y);
            for (int j = 1; j < routes[i].length; j++) {
                int nx = points[routes[i][j] - 1][0] - 1;
                int ny = points[routes[i][j] - 1][1] - 1;
                r.path.add(new int[]{nx, ny});
            }
            robots.add(r);
        }

        // 출발 전 충돌 감지
        Map<List<Integer>, Integer> visit = new HashMap<>();
        for (Robot r : robots) {
            List<Integer> key = Arrays.asList(r.x, r.y);
            visit.put(key, visit.getOrDefault(key, 0) + 1);
        }
        for (int v : visit.values()) {
            if (v > 1) ans += 1;
        }

        // 시뮬레이션
        while (true) {
            boolean flag = true;
            for (Robot r : robots) {
                if (!r.goal) { flag = false; break; }
            }
            if (flag) break;

            List<Robot> moved = new ArrayList<>();

            // 한 칸씩 이동
            for (Robot r : robots) {
                if (r.goal) continue;
                int tx = r.path.get(0)[0];
                int ty = r.path.get(0)[1];
                int curDist = Math.abs(tx - r.x) + Math.abs(ty - r.y);

                for (int[] d : dir) {
                    int nx = r.x + d[0];
                    int ny = r.y + d[1];
                    int nxtDist = Math.abs(tx - nx) + Math.abs(ty - ny);
                    if (0 <= nx && 0 <= ny && nxtDist < curDist) {
                        r.x = nx;
                        r.y = ny;
                        moved.add(r);
                        break;
                    }
                }
            }

            // 이동한 로봇들의 충돌/도착 처리
            Map<List<Integer>, Integer> v = new HashMap<>();
            for (Robot r : moved) {
                List<Integer> key = Arrays.asList(r.x, r.y);
                v.put(key, v.getOrDefault(key, 0) + 1);

                int tx = r.path.get(0)[0];
                int ty = r.path.get(0)[1];

                if (tx == r.x && ty == r.y) {
                    r.path.remove(0);        // 다음 목적지로
                    if (r.path.isEmpty()) {  // 더 이상 목적지 없으면 완료
                        r.goal = true;
                    }
                }
            }

            // 충돌 집계
            for (int va : v.values()) {
                if (va > 1) ans += 1;
            }
        }

        return ans;
    }
}

File: 프로그래머스/2/340211. ［PCCP 기출문제］ 3번 ／ 충돌위험 찾기/［PCCP 기출문제］ 3번 ／ 충돌위험 찾기.kt
import java.util.*;
import kotlin.math.*;
data class Robot(
    var x: Int,
    var y: Int,
    val path: ArrayDeque<Pair<Int,Int>>,
    var goal: Boolean
)

class Solution {
    val directions = listOf(Pair(1,0),Pair(-1,0),Pair(0,1),Pair(0,-1))
    fun solution(points: Array<IntArray>, routes: Array<IntArray>): Int {
        var ans: Int = 0
        val n = points.size
        val x = routes.size
        val m = routes[0].size
        // 로봇은 이동 시 항상 최단 경로, r -> c 순으로 이동
        // 남은 로봇
        val Robots = mutableListOf<Robot>()
        // 시작부터 겹치는 거 판단
        val start = mutableMapOf<Pair<Int,Int>,Int>()
        for ((i,r) in routes.withIndex()) {
            val num = i+1
            val start_point = r[0] - 1
            val (sx,sy) = points[start_point]
            val q = ArrayDeque<Pair<Int,Int>>()
            for (j in 1 until m) {
                val p = r[j] - 1
                val (x,y) = points[p]
                q.addLast(Pair(x-1,y-1))
            }
            start[Pair(sx-1,sy-1)] = start.getOrDefault(Pair(sx-1,sy-1),0) + 1
            val robot = Robot(sx-1,sy-1,q,false)
            Robots.add(robot)
        }
        for (v in start.values) {
            if (v > 1) ans += 1
        }
        
        while (true) {
            var flag = true
            for (r in Robots) {
                if (!r.goal) {
                    flag = false
                    break
                }
            }
            if (flag) break
            
            // 이동한 로봇만 충돌 감지
            val moved = mutableListOf<Robot>()
            for (robot in Robots) {
                if (robot.goal) continue
                val (cx,cy,path,goal) = robot
                val (tx,ty) = path.peekFirst()
                val curD = abs(cx - tx) + abs(cy - ty)
                for ((dx,dy) in directions) {
                    val nx = cx + dx
                    val ny = cy + dy
                    // 거리가 줄어들어야 함
                    val nd = abs(nx - tx) + abs(ny - ty)
                    if (0 <= nx && 0 <= ny && nd < curD) {
                        robot.x = nx
                        robot.y = ny
                        moved.add(robot)
                        break
                    }
                }
            } 
            // 이동한 로봇에 대해서 체크
            // 새로 이동한 곳 충돌 체크 및 도착 체크
            val v = mutableMapOf<Pair<Int,Int>,Int>()
            for (robot in moved) {
                val (cx,cy,path,goal) = robot
                val (tx,ty) = path.peekFirst()
                if (cx == tx && cy == ty) {
                    path.pollFirst()
                    if (path.isEmpty()) {
                        robot.goal = true
                    }
                }
                val p = Pair(cx,cy)
                v[p] = v.getOrDefault(p,0) + 1
            }
            
            for (cnt in v.values) {
                if (cnt > 1) ans++
            }
        }
        
        return ans
    }
}